File: app/src/main/java/com/pr0gramm/app/ui/RequestPasswordRecoveryActivity.java
package com.pr0gramm.app.ui;

import android.os.Bundle;
import android.support.v4.util.PatternsCompat;
import android.widget.Button;
import android.widget.EditText;

import com.jakewharton.rxbinding.widget.RxTextView;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.services.UserService;
import com.pr0gramm.app.ui.base.BaseAppCompatActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;

import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;

public class RequestPasswordRecoveryActivity extends BaseAppCompatActivity {
    @BindView(R.id.submit)
    Button submit;

    @BindView(R.id.email)
    EditText email;

    @Inject
    UserService userService;

    @Override
    protected void injectComponent(ActivityComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(ThemeHelper.theme().basic);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_request_password_recovery);

        RxTextView.textChanges(email)
                .compose(bindToLifecycle())
                .map(val -> String.valueOf(val).trim())
                .map(val -> PatternsCompat.EMAIL_ADDRESS.matcher(val).matches())
                .subscribe(submit::setEnabled);
    }

    @OnClick(R.id.submit)
    public void submitButtonClicked() {
        String email = String.valueOf(this.email.getText()).trim();
        userService.requestPasswordRecovery(email)
                .compose(bindToLifecycleAsync().forCompletable())
                .subscribe(this::requestCompleted, defaultOnError());
    }

    private void requestCompleted() {
        DialogBuilder.start(this)
                .content(R.string.request_password_recovery_popup_hint)
                .positive(R.string.okay, d -> finish())
                .show();
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/RequestPasswordRecoveryActivity.kt
package com.pr0gramm.app.ui

import android.os.Bundle
import androidx.core.util.PatternsCompat
import com.pr0gramm.app.R
import com.pr0gramm.app.databinding.ActivityRequestPasswordRecoveryBinding
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.ui.base.BaseAppCompatActivity
import com.pr0gramm.app.ui.base.bindViews
import com.pr0gramm.app.ui.base.launchWhenStarted
import com.pr0gramm.app.ui.base.withViewDisabled
import com.pr0gramm.app.util.addTextChangedListener
import com.pr0gramm.app.util.di.instance
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext

class RequestPasswordRecoveryActivity : BaseAppCompatActivity("RequestPasswordRecoveryActivity") {
    private val views by bindViews(ActivityRequestPasswordRecoveryBinding::inflate)

    private val userService: UserService by instance()

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(ThemeHelper.theme.basic)

        super.onCreate(savedInstanceState)

        setContentView(views)

        views.email.addTextChangedListener { changedText ->
            val valid = PatternsCompat.EMAIL_ADDRESS.matcher(changedText.trim()).matches()
            views.submit.isEnabled = valid
        }

        views.submit.setOnClickListener {
            submitButtonClicked()
        }
    }

    private fun submitButtonClicked() {
        val email = views.email.text.toString().trim()

        launchWhenStarted(busyIndicator = true) {
            withViewDisabled(views.submit) {
                withContext(NonCancellable + Dispatchers.Default) {
                    userService.requestPasswordRecovery(email)
                }
            }

            requestCompleted()
        }
    }

    private fun requestCompleted() {
        showDialog(this) {
            content(R.string.request_password_recovery_popup_hint)
            positive(R.string.okay) { finish() }
        }
    }
}
